subsys/debug/tracing/cpu_stats.c

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <tracing_cpu_stats.h>
+
+enum cpu_state {
+	CPU_STATE_IDLE,
+	CPU_STATE_NON_IDLE,
+	CPU_STATE_SCHEDULER
+};
+
+static enum cpu_state last_cpu_state = CPU_STATE_SCHEDULER;
+
+static volatile u32_t last_time;
+
+static volatile u64_t idle_cnt;
+static volatile u64_t non_idle_cnt;
+static volatile u64_t sched_cnt;
+
+static struct k_thread *current_thread;
+
+#ifndef CONFIG_SMP
+extern k_tid_t const _idle_thread;
+#endif
+
+static inline int is_idle_thread(struct k_thread *thread)
+{
+#ifdef CONFIG_SMP
+	return thread->base.is_idle;
+#else
+	return thread == _idle_thread;
+#endif
+}
+
+void increase_counter(volatile u64_t *cnt)
+{
+	u32_t time = k_cycle_get_32();
+
+	if (time >= last_time) {
+		(*cnt) += (time - last_time);
+	} else {
+		(*cnt) += (UINT32_MAX - last_time + 1 + time);
+	}
+	last_time = time;
+}
+
+static void cpu_meas_update_counters(void)
+{
+	switch (last_cpu_state) {
+	case CPU_STATE_IDLE:
+		increase_counter(&idle_cnt);
+	break;
+
+	case CPU_STATE_NON_IDLE:
+		increase_counter(&non_idle_cnt);
+	break;
+
+	case CPU_STATE_SCHEDULER:
+		increase_counter(&sched_cnt);
+	break;
+
+	default:
+		__ASSERT_NO_MSG(false);
+	break;
+	}
+}
+
+u32_t cpu_meas_sched_get(void)
+{
+	cpu_meas_update_counters();
+	return SYS_CLOCK_HW_CYCLES_TO_NS(sched_cnt) / 1000000;
```

Comment:
Windowing is applied by resetting counters periodically or calling function from API (`cpu_meas_reset_counters`) - I will change commit message. 

Division here is to change ns to ms - so result will be more human readable (i will add comment to clarify this one). SYS_CLOCK_HW_CYCLES_TO_NS (x) is macro from Zephyr kernel API which gives us result time in ns - so I think it is portable. 
Mentioned divisions are performed rather rarely (once per providing result to user), so maybe it's not a problem.

`cpu_meas_sched_get` is part of API for user (same as `cpu_meas_idle_get` and `cpu_meas_non_idle_get`). I wanted to make user able to use code in two ways: delayed work printing percent of CPU usage periodically (with configurable period) or using API - for example to measure CPU usage in some important section for the code (user can reset counters, make measurement and then get results in ms for scheduler, idle thread, non idle threads). I think it may be convenient for users.

Reply:
@MarekPieta you could change the api to return times in ns (e.g. cpu_stats_sched_get_ns) and let the use do the division if needed. I know we use it on target but it may be that somebody would simply pass the values outside of the target.
Unless you would like to make sure we fit in u32_t... (maybe this should be extended too then)